lib/buffy/throttle_and_timed.ex

```diff
@@ -0,0 +1,381 @@
+
+      @doc false
+      @impl GenServer
+      @spec init({ThrottleAndTimed.key(), ThrottleAndTimed.args()}) :: {:ok, Buffy.ThrottleAndTimed.state()}
+      def init({key, args}) do
+        {:ok, schedule_throttle_and_update_state(%{key: key, args: args, timer_ref: nil})}
+      end
+
+      @doc """
+      Function to invoke the throttle logic if process already exists.
+      It will only schedule a throttle if `timer_ref` is `nil`.
+
+      """
+      @impl GenServer
+      @spec handle_cast(:throttle, Buffy.ThrottleAndTimed.state()) :: {:noreply, Buffy.ThrottleAndTimed.state()}
+      def handle_cast(:throttle, %{timer_ref: nil} = state) do
+        {:noreply, schedule_throttle_and_update_state(state)}
+      end
+
+      def handle_cast(:throttle, state) do
+        {:noreply, state}
+      end
+
+      defp schedule_throttle_and_update_state(state) do
+        timer_ref = Process.send_after(self(), :execute_throttle_callback, unquote(throttle))
+        %{state | timer_ref: timer_ref}
+      end
+
+      @doc false
+      @impl GenServer
+      @spec handle_info(:timeout | :execute_throttle_callback, Buffy.ThrottleAndTimed.state()) ::
+              {:noreply, Buffy.ThrottleAndTimed.state(), {:continue, :do_work}}
+      def handle_info(:timeout, %{key: key, args: args} = state) do
+        :telemetry.execute(
+          [:buffy, :throttle, :timeout],
+          %{count: 1},
+          %{args: args, key: key, module: __MODULE__}
+        )
+
+        {:noreply, state, {:continue, :do_work}}
+      end
+
+      def handle_info(:execute_throttle_callback, state) do
+        {:noreply, state, {:continue, :do_work}}
+      end
+
+      @doc false
+      @impl GenServer
+      @spec handle_continue(do_work :: atom(), Buffy.ThrottleAndTimed.state()) ::
+              {:noreply, Buffy.ThrottleAndTimed.state()} | {:noreply, Buffy.ThrottleAndTimed.state(), timeout()}
+      def handle_continue(:do_work, %{key: key, args: args} = state) do
+        :telemetry.span(
+          [:buffy, :throttle, :handle],
+          %{args: args, key: key, module: __MODULE__},
+          fn ->
+            result = handle_throttle(args)
+            {result, %{args: args, key: key, module: __MODULE__, result: result}}
+          end
+        )
+
+        new_state = %{state | timer_ref: nil}
+        maybe_add_inbox_timeout({:noreply, new_state})
+      rescue
+        e ->
+          Logger.error("Error in throttle: #{inspect(e)}")
+          new_state = %{state | timer_ref: nil}
+          maybe_add_inbox_timeout({:noreply, new_state})
+      end
+
+      defp maybe_add_inbox_timeout({return_signal, state} = return_tuple) do
+        loop_interval = unquote(loop_interval)
+
+        if is_number(loop_interval) do
+          {return_signal, state, loop_interval}
+        else
+          Logger.error(
+            "Error parsing :loop_interval - value is not a number, will ignore. Got: #{inspect(loop_interval)}"
+          )
+
+          return_tuple
+        end
```

Comment:
Should we move this check up, potentially to the top of the macro when we use `Keyword.fetch!/2` to get the value so that we ensure before we even start the process that the value of `looop_interval` is a number?

If we can't do it outside the `defmacro`, maybe we could do it inside the `start_link`?

What do you think?

Reply:
Ooh yes of course, why run it when the field type is wrong? We should let it crash immediately :) I'll make the changes!

Reply:
Done in 16be37c